src/stirling/source_connectors/socket_tracer/protocols/mqtt/stitcher_test.cc

```diff
@@ -0,0 +1,827 @@
+  pub1_pid3_frame.timestamp_ns = ++t;
+  puback_frame.timestamp_ns = ++t;
+  pubrec_pid2_frame.timestamp_ns = ++t;
+  pubrel_pid2_frame.timestamp_ns = ++t;
+  pubcomp_pid2_frame.timestamp_ns = ++t;
+  puback_pid3_frame.timestamp_ns = ++t;
+  req_map[1].push_back(pub1_frame);
+  req_map[2].push_back(pub2_pid2_frame);
+  req_map[3].push_back(pub1_pid3_frame);
+  req_map[2].push_back(pubrel_pid2_frame);
+  resp_map[1].push_back(puback_frame);
+  resp_map[2].push_back(pubrec_pid2_frame);
+  resp_map[2].push_back(pubcomp_pid2_frame);
+  resp_map[3].push_back(puback_pid3_frame);
+  result = StitchFrames(&req_map, &resp_map);
+  EXPECT_TRUE(AreAllDequesEmpty(resp_map));
+  EXPECT_EQ(TotalDequeSize(req_map), 0);
+  EXPECT_EQ(result.error_count, 0);
+  EXPECT_EQ(result.records.size(), 4);
+  req_map.clear();
+  resp_map.clear();
+}
+
+TEST(MqttStitcherTest, DummyResponseStitching) {
+  // Test for requests that do not have responses
+  absl::flat_hash_map<packet_id_t, std::deque<Message>> req_map;
+  absl::flat_hash_map<packet_id_t, std::deque<Message>> resp_map;
+
+  std::string_view frame_view;
+  RecordsWithErrorCount<Record> result;
+
+  Message pub0_frame, disconnect_frame, auth_frame, connect_frame, connack_frame;
+  frame_view = CreateStringView<char>(CharArrayStringView<uint8_t>(kPublishQosZero));
+  ParseFrame(message_type_t::kRequest, &frame_view, &pub0_frame);
+
+  frame_view = CreateStringView<char>(CharArrayStringView<uint8_t>(kDisconnectFrame));
+  ParseFrame(message_type_t::kRequest, &frame_view, &disconnect_frame);
+
+  frame_view = CreateStringView<char>(CharArrayStringView<uint8_t>(kAuthFrame_success));
+  ParseFrame(message_type_t::kRequest, &frame_view, &auth_frame);
+
+  frame_view = CreateStringView<char>(CharArrayStringView<uint8_t>(kConnect));
+  ParseFrame(message_type_t::kRequest, &frame_view, &connect_frame);
+
+  frame_view = CreateStringView<char>(CharArrayStringView<uint8_t>(kConnack));
+  ParseFrame(message_type_t::kResponse, &frame_view, &connack_frame);
+
+  int t = 0;
+  pub0_frame.timestamp_ns = ++t;
+  disconnect_frame.timestamp_ns = ++t;
+  auth_frame.timestamp_ns = ++t;
+  req_map[0].push_back(pub0_frame);
+  req_map[0].push_back(disconnect_frame);
+  req_map[0].push_back(auth_frame);
+  result = StitchFrames(&req_map, &resp_map);
+  EXPECT_EQ(TotalDequeSize(req_map), 0);
+  EXPECT_EQ(result.error_count, 0);
+  EXPECT_EQ(result.records.size(), 3);
+
+  t = 0;
+  pub0_frame.timestamp_ns = ++t;
+  connect_frame.timestamp_ns = ++t;
+  disconnect_frame.timestamp_ns = ++t;
+  auth_frame.timestamp_ns = ++t;
+  connack_frame.timestamp_ns = ++t;
+  req_map[0].push_back(pub0_frame);
+  req_map[0].push_back(connect_frame);
+  req_map[0].push_back(disconnect_frame);
+  req_map[0].push_back(auth_frame);
+  resp_map[0].push_back(connack_frame);
+  result = StitchFrames(&req_map, &resp_map);
+  EXPECT_EQ(TotalDequeSize(req_map), 0);
+  EXPECT_EQ(result.error_count, 0);
+  EXPECT_EQ(result.records.size(), 4);
+}
+
```

Comment:
> Duplicate frame during a QoS 1 PUBLISH

I am assuming you mean a duplicate QOS 1 PUBLISH from the client itself, with a single PUBACK from the server (caused due to retransmission of PUBLISH after not receiving PUBACK). This pointed to a bug in the stitcher code where I was not checking if the response had been consumed in the response deque in the code where I was searching for a response with the matching response key. This was causing a single response to be matched to multiple duplicate requests. I have added this, and now the code works like this in case of duplicate QOS 1 PUBLISH,
1. If there are two duplicate PUBLISH frames (QOS 1) in the request deque, and one PUBACK frame in the response deque, then the first PUBLISH frame is stitched with the PUBACK frame, and the second PUBLISH frame is left in the request deque (with error count as 0). The code works like this because I did not want to clear certain request frames in the end of the deque, as the response could be filled the next time StitchFrames is called and the response deque now contains the response to this request. We know however that no response will come for this duplicate PUBLISH request, but I see no problem with this as this request will be cleared in the next round when more request-response stitching occurs in the deque.
2. If there are two duplicate QOS 1 PUBLISH frames, one SUBSCRIBE frame (can be any other request frame) in the request deque, and one PUBACK frame, one SUBACK frame in the response deque, then the first PUBLISH will be stitched with PUBACK, the SUBSCRIBE will be stitched with SUBACK, and this time the error count will be 1, and the duplicate PUBLISH will be cleared from the request deque.

> Case were client and server are using the same packet identifiers -- this is possible according to the spec and is different from how other protocols that use the StitchFrames map interface work (ID generation is usually client side only).

This would work fine if the flip-flopping as you mentioned does not occur, but if it does, then I will have to add logic to do reverse stitching from response to request as well, as you said. I will add a TODO comment for reference in the future.

Reply:
For case 1, how does the implementation discern between an unanswered `PUBLISH` frame (QOS 1) compared to a `PUBLISH` sequence (QOS1) where there are duplicated `PUBLISH`es and a `PUBACK`? It seems that the next round of stitching might have difficulty telling the two cases apart. Imo the former should be classified as an error while the latter seems fine to consider as a successful case.

Sounds good on the TODO and revisiting this for case 2.

Reply:
> how does the implementation discern between an unanswered PUBLISH frame (QOS 1) compared to a PUBLISH sequence (QOS1) where there are duplicated PUBLISHes and a PUBACK? 

Actually the implementation would consider one of the duplicated PUBLISH frames (QOS 1) (or multiple, if there are more than one duplicate frames due to retransmission) as unanswered PUBLISH frames. In the next round of stitching, the unanswered PUBLISH frames would be discarded, when newer frames which have newer responses are stitched.
That logic is present in this code,
```
while (iter != req_deque.end() && (iter->timestamp_ns < latest_resp_ts)) {
  if (iter->consumed) {
    ++erase_until_iter;
  }
  if (!iter->consumed && !(iter == req_deque.end() - 1) && ((erase_until_iter + 1)->consumed)) {
    ++error_count;
    ++erase_until_iter;
  }
  ++iter;
}
req_deque.erase(req_deque.begin(), erase_until_iter);
}
```
where all the requests that have not been consumed until the latest response timestamp would be removed. I thought this would make sense, since if there are responses to subsequent requests after the unanswered PUBLISH, then it will go unanswered and I need to remove it from the request deque.

But yeah, it will give error count as 1 in cases where are there duplicate PUBLISH frames with a PUBACK. I thought this was fine since it is a case of an unanswered PUBLISH itself that prompted retransmission and duplication, so error count being the number of duplicate PUBLISH frames minus one (for the answered PUBLISH) would make sense.

Should I modify the logic so that the error count is zero in cases where the PUBLISH frames finally has an answer in PUBACK, and error count is non  zero in cases where there are no PUBACKs at all?

Reply:
> Should I modify the logic so that the error count is zero in cases where the PUBLISH frames finally has an answer in PUBACK, and error count is non zero in cases where there are no PUBACKs at all?

Yes, I believe that's the correct thing to do here. Stirling will disable tracing a connection if the stitcher failure rate goes over a certain threshold, so accurately labeling these cases will avoid difficult to debug situations later on.

Reply:
I was thinking, the way I could do this is marking the duplicate messages that have been answered as consumed.

Also, the stitching logic would match the PUBLISH with `dup` bit set to 0 with the PUBACK, and the PUBLISH with `dup` bit set to 1 would be dropped. Is this fine? Should I replace the PUBLISH in the entries vector once a duplicate has been found that has already been stitched with a PUBACK with the first PUBLISH frame?

Reply:
> Also, the stitching logic would match the PUBLISH with dup bit set to 0 with the PUBACK, and the PUBLISH with dup bit set to 1 would be dropped. Is this fine?

It feels like the response (PUBREC, etc) should ideally be matched against the latest dup. Matching against the non duplicate frame gives the false impression that the duplicate shouldn't be expected (meaning the original message was delivered properly).

Maybe we can revisit this thread after you have a chance to consider my comments about using state to handle the duplicates.

Reply:
Yeah using state to handle duplicates makes this really simple, now the latest duplicate can be matched to the response.